src/utils/index.ts

```diff
@@ -1,57 +1,372 @@
+  quantity: BigInt,
+  redeemOrTransferAmounts: RedeemOrTransferAmounts | null
+): void {
+  // get tokens
+  const underlyingToken = fetchToken(optionType.underlyingAsset);
+  const exerciseToken = fetchToken(optionType.exerciseAsset);
+
+  // get asset market prices for metrics
+  const underlyingPriceUSD = getTokenPriceUSD(underlyingToken.id);
+  const exercisePriceUSD = getTokenPriceUSD(exerciseToken.id);
+
+  // calculate total token amounts for notional values
+  let underlyingAmountTotal = BigInt.fromI32(0);
+  let exerciseAmountTotal = BigInt.fromI32(0);
+  if (eventKind == "write") {
+    underlyingAmountTotal = optionType.underlyingAmount.times(quantity);
+    exerciseAmountTotal = optionType.exerciseAmount.times(BigInt.fromI32(0));
+  }
+  if (eventKind == "exercise") {
+    underlyingAmountTotal = optionType.underlyingAmount.times(
+      BigInt.fromI32(0)
+    );
+    exerciseAmountTotal = optionType.exerciseAmount.times(quantity);
+  }
+  if (eventKind == "redeem" || eventKind == "transfer") {
+    underlyingAmountTotal = redeemOrTransferAmounts!.underlyingAmountTotal;
+    exerciseAmountTotal = redeemOrTransferAmounts!.exerciseAmountTotal;
+  }
+
+  // get prices in USD
+  const underlyingTotalUSD = underlyingPriceUSD
+    .times(underlyingAmountTotal.toBigDecimal())
+    .div(exponentToBigDecimal(BigInt.fromI64(underlyingToken.decimals)));
+  const exerciseTotalUSD = exercisePriceUSD
+    .times(exerciseAmountTotal.toBigDecimal())
+    .div(exponentToBigDecimal(BigInt.fromI64(exerciseToken.decimals)));
+
+  // get daily metrics
+  const dailyOSEMetrics = fetchDailyOSEMetrics(timestamp);
+  const underlyingDaily = fetchDailyTokenMetrics(underlyingToken.id, timestamp);
+  const exerciseDaily = fetchDailyTokenMetrics(exerciseToken.id, timestamp);
+
+  // save previous TVLs
+  const dailyTVLUSDBefore = dailyOSEMetrics.totalValueLockedUSD;
+  const underlyingTVLUSDBefore = underlyingDaily.totalValueLockedUSD;
+  const exerciseTVLUSDBefore = exerciseDaily.totalValueLockedUSD;
+
+  // calculate new TVLs
+  const underlyingTVLUSDAfter = underlyingPriceUSD.times(
+    underlyingToken.totalValueLocked
+      .toBigDecimal()
+      .div(exponentToBigDecimal(BigInt.fromI64(underlyingToken.decimals)))
+  );
+  const exerciseTVLUSDAfter = exercisePriceUSD.times(
+    exerciseToken.totalValueLocked
+      .toBigDecimal()
+      .div(exponentToBigDecimal(BigInt.fromI64(exerciseToken.decimals)))
+  );
 
-    feeSwitch = new FeeSwitch(contractAddress);
-    feeSwitch.feeToAddress = initialFeeToAddress;
-    feeSwitch.isEnabled = false;
-    feeSwitch.save();
+  /**
+   *  Update Daily OSE Metrics
+   */
+  // By saving the before/after TVL in USD, we are able to maintain consistent records
+  dailyOSEMetrics.totalValueLockedUSD = dailyOSEMetrics.totalValueLockedUSD
+    .minus(underlyingTVLUSDBefore)
+    .minus(exerciseTVLUSDBefore)
+    .plus(underlyingTVLUSDAfter)
+    .plus(exerciseTVLUSDAfter);
+
+  if (eventKind == "write") {
```

Comment:
Not entirely sure how this flow control works, but could it be more efficient to use if/elseif/else block?

Reply:
Ideally, I would use a switch statement, but that's not a thing in current state of AssemblyScript. I'll refactor to use if/else's though.

Reply:
elifs added https://github.com/valorem-labs-inc/valorem-subgraph/commit/0c078e644a7c38e9ff7af689a1784ff579a6112f